[TheOneric] Enable strict mode in js

dmitrylyzo opened the pull request:
_Pulling from the upstream._
_Original author: TheOneric_

https://github.com/libass/JavascriptSubtitlesOctopus/pull/143

TheOneric committed:
Enable strict mode in js

This will make some types of bugs easier to detect
and avoid accidental additions of code not working
inside Modules, which always use strict mode.

Performance measurements show no relevant performance
difference between strict and non-strict mode, so this
won't cause regressions in this regard.

Cherry-picked from: https://github.com/libass/JavascriptSubtitlesOctopus/commit/c8f02d664b86e68fdafd69e53cc87ed2ede413f1

TheOneric committed:
disable closure

There's now a new warning when closure isn't set explicitly
for non-WASM builds. Currently, there are errors concerning
redeclaration of the variable screen with closure enabled,
so explicitly disable it.
For the future, we probably want to remove the redeclaration
and enable closure.

Cherry-picked from: https://github.com/libass/JavascriptSubtitlesOctopus/commit/45a05dba22c16d85ea5a2d38633defc14777303a

TheOneric committed:
fix EVAL_CTORS

EVAL_CTORS now errors when used without WASM,
previosuly it was likely just silently disabled.
Thus, only set it when linking with WASM.

Cherry-picked from: https://github.com/libass/JavascriptSubtitlesOctopus/commit/45a05dba22c16d85ea5a2d38633defc14777303a

dmitrylyzo commented:
This doesn't seem to add "use strict;" to workers - `null;` has been added instead.
Emscripten 3.1.24 adds "use strict;" to the normal worker, but not to the legacy one.

UPD:
Emscripten 3.1.25 adds "use strict;" to both workers. So this the minimum version for the STRICT_JS feature.

TheOneric commented:
> Emscripten 3.1.25 adds "use strict;" to both workers. So this the minimum version for the STRICT_JS feature.

`STRICT_JS` existed since emscripten 1.38.43 (also see [introducing commit](https://github.com/emscripten-core/emscripten/commit/2fe944f72f926c6c5b9dca4626ac2d5cb27f9c0e)). So it was *supposed* to work for much longer, however as you noticed on some versions it didn't, good catch.
It turns out an optimiser pass of emscripten turned `"use strict";` int `null;`, which was [fixed](https://github.com/emscripten-core/emscripten/issues/17963) with emscripten 3.1.24 *(not sure what changed with 3.1.25 or when this bug was first introduced)*.

dmitrylyzo committed:
Update emscripten to 3.1.24

It turns out an optimizer pass of emscripten turned `"use strict";`
into `null;`, which was fixed with emscripten 3.1.24

dmitrylyzo committed:
Re-add calledMain variable using Module[postRun]

`calledMain` global variable was removed from the JS runtime
in emscripten 3.1.17
